mapillary_tools/download_blurred.py

```diff
@@ -0,0 +1,413 @@
+    try:
+        for i in range(number_threads):
+            t = BlurredOriginalsDownloader(
+                lock,
+                q,
+                output_folder,
+                headers,
+                counter,
+                done_file_obj,
+                err_file_obj)
+            threads.append(t)
+            t.start()
+
+        monitor = WorkerMonitor(threads, q)
+        monitor.start()
+
+        while True:
+            any_alive = False or monitor.is_alive()
+
+            for t in threads:
+                any_alive = (any_alive or t.is_alive())
+
+            if not any_alive:
+                break
+            time.sleep(0.5)
+
+    except (ServiceExit, KeyboardInterrupt):
+        print('\nexit or interrupt\n')
+        monitor.shutdown_flag.set()
+
+        for t in threads:
+            t.shutdown_flag.set()
+
+        for t in threads:
+            t.join()
+            print('INFO: Thread stopped {}'.format(t))
+
+        monitor.join()
+        print('INFO: Thread stopped {}'.format(monitor))
+
+        all_file_obj.close()
+        done_file_obj.close()
+        err_file_obj.close()
+        sys.exit(1)
+
+
+class BlurredOriginalsDownloader(threading.Thread):
+    """
+    """
+
+    def __init__(self,
+                 lock,
+                 q,
+                 output_folder,
+                 headers,
+                 counter,
+                 done_file,
+                 err_file):
+        threading.Thread.__init__(self)
+
+        self.counter = counter
+        self.done_file = done_file
+        self.err_file = err_file
+        self.headers = headers
+        self.queue = q
+        self.lock = lock
+        self.output_folder = output_folder
+        self.shutdown_flag = threading.Event()
+        self.worker_stats = 0
+
+    def download_file(self, image_key, filename):
+        download_url = "{}/v3/images/{}/download_original?client_id={}".format(
+            MAPILLARY_ENDPOINT,
+            image_key,
+            uploader.CLIENT_ID)
+
+        response = requests.get(
+            download_url, stream=True, headers=self.headers)
+
+        if response.status_code != 200:
+            print(response.json())
+            return False
+
+        with open(filename, "wb") as f:
+            total_length = response.headers.get('content-length')
+
+            dl = 0
+            total_length = int(total_length)
+            for data in response.iter_content(chunk_size=4096):
+                dl += len(data)
+                f.write(data)
+        return True
+
+    def run(self):
+        while not self.shutdown_flag.is_set():
+            self.lock.acquire()
+            if (self.queue.empty()):
+                self.lock.release()
+                break
+
+            current_image_key = self.queue.get_nowait()
+
+            image_path = os.path.join(
+                self.output_folder, '{}.jpg'.format(current_image_key))
+
+            if not os.path.exists(os.path.dirname(image_path)):
```

Comment:
you dont need the ˙create_dirs`  if you have this, or the other way around?

Reply:
I'm moving this to `created_dirs` function.